src/parse/text/simple_extensions/types.rs

```diff
@@ -0,0 +1,1773 @@
+                    parsed_field_type.visit_references(&mut link);
+                    let field_concrete_type = ConcreteType::try_from(parsed_field_type)?;
+
+                    field_types.push(field_concrete_type);
+                }
+
+                Ok(ConcreteType {
+                    kind: ConcreteTypeKind::NamedStruct {
+                        field_names,
+                        field_types,
+                    },
+                    nullable: false,
+                })
+            }
+        }
+    }
+}
+
+/// Invalid type name error
+#[derive(Debug, Error, PartialEq)]
+#[error("invalid type name `{0}`")]
+pub struct InvalidTypeName(String);
+
+/// Known Substrait types (builtin + extension references).
+///
+/// Note that this is a recursive type - other than the [BuiltinType]s, the other variants can
+/// have type parameters that are themselves [ConcreteType]s.
+#[derive(Clone, Debug, PartialEq)]
+pub enum ConcreteTypeKind {
+    /// Built-in Substrait type (primitive or parameterized)
+    Builtin(BuiltinType),
+    /// Extension type with optional parameters
+    Extension {
+        /// Extension type name
+        name: String,
+        /// Type parameters
+        parameters: Vec<TypeParameter>,
+    },
+    /// List type with element type
+    List(Box<ConcreteType>),
+    /// Map type with key and value types
+    Map {
+        /// Key type
+        key: Box<ConcreteType>,
+        /// Value type
+        value: Box<ConcreteType>,
+    },
+    /// Struct type (ordered fields without names)
+    Struct(Vec<ConcreteType>),
+    /// Named struct type (nstruct - ordered fields with names)
+    NamedStruct {
+        /// Field names
+        field_names: Vec<String>,
+        /// Field types (same order as field_names)
+        field_types: Vec<ConcreteType>,
+    },
+}
+
+impl fmt::Display for ConcreteTypeKind {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ConcreteTypeKind::Builtin(b) => write!(f, "{b}"),
+            ConcreteTypeKind::Extension { name, parameters } => {
+                write!(f, "{name}")?;
+                write_separated(f, parameters.iter(), "<", ">", ", ")
+            }
+            ConcreteTypeKind::List(elem) => write!(f, "list<{elem}>"),
+            ConcreteTypeKind::Map { key, value } => write!(f, "map<{key}, {value}>"),
+            ConcreteTypeKind::Struct(types) => {
+                write_separated(f, types.iter(), "struct<", ">", ", ")
+            }
+            ConcreteTypeKind::NamedStruct {
+                field_names,
+                field_types,
+            } => {
+                let kvs = field_names
+                    .iter()
+                    .zip(field_types.iter())
+                    .map(|(k, v)| KeyValueDisplay(k, v, ": "));
+
+                write_separated(f, kvs, "{", "}", ", ")
+            }
+        }
+    }
+}
+
+/// A concrete, fully-resolved type instance
```

Comment:
And then in the spirit of the other comment, lets just make clear why this is distinct from `ConcreteTypeKind`. 
```suggestion
/// A concrete, fully-resolved type instance with nullability.
```